src/compile/invalid/datasources.ts

```diff
@@ -0,0 +1,63 @@
+import {MarkInvalidDataMode} from '../../invalid';
+
+import {DataSourceType} from '../../data';
+import {normalizeInvalidDataMode} from './normalizeInvalidDataMode';
+
+type PreOrPostFilteringInvalidValues = 'pre-filter' | 'post-filter';
+
+export interface DataSourcesForHandlingInvalidValues {
+  marks: PreOrPostFilteringInvalidValues;
+  scales: PreOrPostFilteringInvalidValues;
+}
+interface GetDataSourcesForHandlingInvalidValuesProps {
+  invalid: MarkInvalidDataMode | null | undefined;
+  isPath: boolean;
+}
+
+export function getDataSourcesForHandlingInvalidValues({
```

Comment:
This method includes one of the main rules for defining how these modes affect data sources for scale / marks.